File: terraform/provider/internal/client/keys.go
package client

import (
	"context"
	"fmt"
)

const KeyKind = "Key"

type KeySpec struct {
	Usage        string         `json:"usage,omitempty"`
	MaterialType string         `json:"materialType,omitempty"`
	DesiredState string         `json:"desiredState,omitempty"`
	KeyData      map[string]any `json:"keyData,omitempty"`
}

type KeyStatus struct {
	State             string `json:"state"`
	KeyDataConfigured bool   `json:"keyDataConfigured"`
}

type Key struct {
	TypeMeta
	Metadata ObjectMetadata `json:"metadata"`
	Spec     KeySpec        `json:"spec"`
	Status   *KeyStatus     `json:"status,omitempty"`
}

type CreateKeyRequest struct {
	TypeMeta
	Metadata ObjectMetadata `json:"metadata"`
	Spec     KeySpec        `json:"spec"`
}

type KeySpecPatch struct {
	DesiredState *string `json:"desiredState,omitempty"`
}

type UpdateKeyRequest struct {
	TypeMeta
	Metadata *ObjectMetadataPatch `json:"metadata"`
	Spec     *KeySpecPatch        `json:"spec"`
}

func (c *Client) CreateKey(ctx context.Context, req CreateKeyRequest) (*Key, error) {
	var key Key
	err := c.post(ctx, "/api/v1/keys", req, &key)
	return &key, err
}

func (c *Client) GetKey(ctx context.Context, id string) (*Key, error) {
	var key Key
	err := c.get(ctx, fmt.Sprintf("/api/v1/keys/%s", id), &key)
	return &key, err
}

func (c *Client) UpdateKey(ctx context.Context, id string, req UpdateKeyRequest) (*Key, error) {
	var key Key
	err := c.patch(ctx, fmt.Sprintf("/api/v1/keys/%s", id), req, &key)
	return &key, err
}

func (c *Client) DeleteKey(ctx context.Context, id string) error {
	return c.delete(ctx, fmt.Sprintf("/api/v1/keys/%s", id))
}

File: sdks/js/src/keys.ts
import { client } from './client';
import {
  MutableResourceMetadata,
  NamespacedCreateMetadata,
  ObjectMetadata,
  ResourceList,
  TypeMeta,
} from './common';

export const KEY_KIND = 'Key' as const;

export enum KeyUsage {
  DATA_ENCRYPTION = 'data_encryption',
}

export enum KeyMaterialType {
  SYMMETRIC = 'symmetric',
  PUBLIC = 'public',
  PRIVATE = 'private',
  EXTERNAL = 'external',
}

export enum KeyState {
  ACTIVE = 'active',
  DISABLED = 'disabled',
}

export type KeyData = Record<string, unknown>;

export interface KeyMetadata extends ObjectMetadata {
  id: string;
  name: string;
  namespace: string;
  createdAt: string;
  updatedAt: string;
}

export interface KeySpec {
  usage: KeyUsage;
  materialType: KeyMaterialType;
  desiredState: KeyState;
  /** Provider configuration is always returned with secret fields redacted. */
  keyData?: KeyData;
}

export interface KeyStatus {
  state: KeyState;
  keyDataConfigured: boolean;
}

export interface Key extends TypeMeta<typeof KEY_KIND> {
  metadata: KeyMetadata;
  spec: KeySpec;
  status: KeyStatus;
}

export interface CreateKeyRequest extends TypeMeta<typeof KEY_KIND> {
  metadata: NamespacedCreateMetadata;
  spec: {
    usage?: KeyUsage;
    materialType?: KeyMaterialType;
    desiredState?: KeyState;
    /** Write-only provider configuration; responses contain only redacted values. */
    keyData: KeyData;
  };
}

export interface UpdateKeyRequest extends TypeMeta<typeof KEY_KIND> {
  metadata: MutableResourceMetadata;
  spec: {
    usage?: KeyUsage;
    materialType?: KeyMaterialType;
    desiredState?: KeyState;
    keyData?: KeyData;
  };
}

export type KeyList = ResourceList<Key>;

export interface ListKeysParams {
  name?: string;
  cursor?: string;
  limit?: number;
  state?: KeyState;
  namespace?: string;
  labelSelector?: string;
  orderBy?: string;
}

export const listKeys = (params?: ListKeysParams) =>
  client.get<KeyList>('/api/v1/keys', { params });

export const createKey = (request: CreateKeyRequest) =>
  client.post<Key>('/api/v1/keys', request);

export const getKey = (id: string) => client.get<Key>(`/api/v1/keys/${id}`);

export const updateKey = (id: string, request: UpdateKeyRequest) =>
  client.patch<Key>(`/api/v1/keys/${id}`, request);

export const deleteKey = (id: string) => client.delete(`/api/v1/keys/${id}`);

export const keys = {
  list: listKeys,
  create: createKey,
  get: getKey,
  update: updateKey,
  delete: deleteKey,
};
